src/tests/unit/tests/electron/platform/android/appium-service-configurator.test.ts

```diff
@@ -0,0 +1,303 @@
+
+        const devices = await testSubject.getConnectedDevices();
+
+        expect(devices.length).toBe(1);
+        expect(devices[0].id).toBe(deviceId);
+        expect(devices[0].isEmulator).toBe(false);
+        expect(devices[0].friendlyName).toBe(deviceModel);
+
+        adbMock.verifyAll();
+    });
+
+    it('getConnectedDevices, 1 emulator and 1 physical device detected', async () => {
+        const expectedEmulators = [{ udid: emulatorId }];
+        const expectedDevices = [{ udid: deviceId }, { udid: emulatorId }];
+        const expectedModels: Array<string> = [emulatorModel, deviceModel];
+        let modelIndex = 0;
+        adbMock
+            .setup(m => m.getConnectedEmulators())
+            .returns(() => expectedEmulators)
+            .verifiable(Times.once());
+        adbMock
+            .setup(m => m.getConnectedDevices())
+            .returns(() => expectedDevices)
+            .verifiable(Times.once());
+        adbMock.setup(m => m.setDeviceId(emulatorId)).verifiable(Times.once());
+        adbMock.setup(m => m.setDeviceId(deviceId)).verifiable(Times.once());
+        adbMock
+            .setup(m => m.getModel())
+            .returns(() => expectedModels[modelIndex++])
+            .verifiable(Times.exactly(2));
+
+        const devices = await testSubject.getConnectedDevices();
+
+        expect(devices.length).toBe(2);
+        expect(devices[0].id).toBe(emulatorId);
+        expect(devices[0].isEmulator).toBe(true);
+        expect(devices[0].friendlyName).toBe(emulatorModel);
+        expect(devices[1].id).toBe(deviceId);
+        expect(devices[1].isEmulator).toBe(false);
+        expect(devices[1].friendlyName).toBe(deviceModel);
+
+        adbMock.verifyAll();
+    });
+
+    it('getPackageInfo, version information not found', async () => {
+        const expectedPackageInfo: PackageInfo = {} as PackageInfo;
+        adbMock.setup(m => m.setDeviceId(emulatorId)).verifiable(Times.once());
+        adbMock
+            .setup(m => m.getPackageInfo(servicePackageName))
+            .returns(() => expectedPackageInfo)
+            .verifiable(Times.once());
+
+        await expect(testSubject.getPackageInfo(emulatorId)).rejects.toThrowError(
+            'Unable to obtain package version information',
+        );
+
+        adbMock.verifyAll();
+    });
+
+    it('getPackageInfo, contains versionCode but no versionName', async () => {
+        const versionCode: number = 500;
+        const expectedPackageInfo: PackageInfo = { versionCode } as PackageInfo;
+        adbMock.setup(m => m.setDeviceId(emulatorId)).verifiable(Times.once());
+        adbMock
+            .setup(m => m.getPackageInfo(servicePackageName))
+            .returns(() => expectedPackageInfo)
+            .verifiable(Times.once());
+
+        const info = await testSubject.getPackageInfo(emulatorId);
+        expect(info.versionCode).toBe(versionCode);
+        expect(info.versionName).toBeUndefined();
```

Comment:
Is there a case where this would be expected? I would have expected us to treat this as an error

Reply:
versionName is optional, but I agree that we should always be setting it

Reply:
I updated this to make versionName and versionCode both optional, so that we can distinguish between "the package isn't installed" and "the package is installed but we can't get useful version info". In the former case, we would just install the package, while in the latter case we would uninstall the existing package before installing the new one